### ISSUE TITLE
Add ArcArray (and replace RcArray?)

### ISSUE THREAD

**Issue by bluss** [bluss/rust-ndarray#400]
I'm thinking of adding ArcArray to use reference counting to help saving memory in the dataset project (which is not published yet).

The thought is.. should we outright replace RcArray by ArcArray, are there use cases where we want to save overhead of atomic ref counting?  The use case for ArcArray would be to have multiple handles to datasets. The benefit of having ArcArray instead of `Arc<Array<T, D>>` is that the former also makes subsets that are sliced in place, share the same reference to the data.

**Comment by bluss** [comment]
cc @jturner314 Do you think it's feasible to just replace RcArray with ArcArray? Just replacing would solve some minor issues with how the conversions work between Array <=> RcArray right now (it doesn't extend very well to a situation where all three exist at the same time).

**Comment by jturner314** [comment]
As a disclaimer, I have read the implementation of `RcArray`, but I've never actually used it in client code. That said, I think it would be fine to replace `RcArray` with `ArcArray`.

The performance difference will be the difference between non-atomic and atomic operations on the reference count. My understanding is that atomic operations are significantly more expensive, but in the case of `ndarray`, I wouldn't expect that to be much of an issue. Afaict, there are three places where an atomic operation would be performed:

1. When an `ArcArray` is cloned.
2. When an `ArcArray` is dropped.
3. When the data in an `ArcArray` is mutated.

There may also be an atomic operation when a new `ArcArray` is created; I'm not sure. I'm not really concerned about making creation/clones/drops more expensive because the user generally tries to minimize the number of those operations anyway. I suppose it could be an issue if someone was creating *a lot* of `RcArray` instances (maybe some sort of recursive algorithm), but that seems like a rare case.

At first glance, it may seem like an issue to make mutation much more expensive, but mutation of an `RcArray` is already expensive anyway in the non-unique case because the data needs to be cloned. If the array uniquely owns the data and you need to mutate it multiple times, you can bypass most of the atomic operations by first converting it to an `Array` with `.to_owned()` (1 atomic op), then performing the mutations (0 atomic ops), then converting back to `ArcArray` if desired (1? atomic op).

In conclusion, the only case where I'd expect removing `RcArray` to be a significant issue would be if someone was creating *a lot* of `RcArray` instances, but that seems like a rare case. A user may also need to modify their code (temporarily convert to `Array`) if they are mutating the same `RcArray` multiple times, but once they do, the performance impact of `ArcArray` shouldn't be too large.

**Comment by bluss** [comment]
I agree with you. The main motivation was that it seemed to fall out of the code: it's much simpler to replace RcArray than to duplicate it. No problem to add another Array implementation but there are some fixed points like `ArrayBase::into_shared()` etc that needed duplication.
